File: week_3/자물쇠와 열쇠.js
Array.prototype.rotate90 = function () {
	const copied = this.map((line) => [...line]);
	copied.forEach((line, y) => {
		line.forEach((_, x) => (this[y][x] = !!copied[this.length - 1 - x][y]));
	});
};

function derive_key(key, controlLock, env) {
	let { count } = env;
	const { m, n, bx, by } = env;
	const gap = m - 1;
	const sx = Math.max(bx, gap);
	const sy = Math.max(by, gap);
	const ex = Math.min(bx + m, gap + n);
	const ey = Math.min(by + m, gap + n);

	for (let y = sy; y < ey; y++) {
		for (let x = sx; x < ex; x++) {
			const key_v = key[y - by][x - bx];
			const lock_v = controlLock[y][x];
			count -= key_v && lock_v;
			if (key_v ^ lock_v) return false;
		}
	}

	return count === 0;
}

function solution(key, lock) {
	const [m, n] = [key.length, lock.length];
	const ext = n + 2 * (m - 1);

	let count = 0;
	const controlLock = lock.reduce(
		(controled, line, y) => {
			line.forEach((value, x) => {
				count += !value;
				controled[y + (m - 1)][x + (m - 1)] = !value;
			});
			return controled;
		},
		Array.from(Array(ext), () => Array(ext).fill(false))
	);

	const compL = n + (m - 1);
	for (let i = 0; i < 4; i++) {
		key.rotate90();
		for (let y = 0; y < compL; y++) {
			for (let x = 0; x < compL; x++) {
				const env = { m, n, bx: x, by: y, count };
				if (derive_key(key, controlLock, env)) return true;
			}
		}
	}

	return false;
}

File: week_3/자물쇠와 열쇠.py
# Key 와 Lock을 비교할 때 수월하기 위해 Lock을 변경
# 1. Lock의 길이를 확장, 0과 1을 반대로 변환
# 2. Lock의 홈 개수 확인
def induce_lock(lock, m):
    n = len(lock)
    ext = n + 2 * (m - 1)
    count = 0
    converted = [[False] * ext for _ in range(ext)]
    for y in range(n):
        for x in range(n):
            count += not lock[y][x]
            converted[y + (m - 1)][x + (m - 1)] = not lock[y][x]
    
    return (converted, count)

def rotate_90_graph(graph):
    l = len(graph)
    result = [[False] * l for _ in range(l)]
    
    for y in range(l):
        for x in range(l):
            result[y][x] = bool(graph[l - 1 - x][y])
    
    return result

# 중앙 Lock 부분만을 비교
def derive_key(key, induced, standard, count):
    m, n = len(key), len(induced) - (2 * (len(key) - 1))
    bx, by = standard
    gap = m - 1
    sx, sy = max(bx, gap), max(by, gap)
    ex, ey = min(bx + m, gap + n), min(by + m, gap + n)

    for y in range(sy, ey):
        for x in range(sx, ex):
            key_v, lock_v = key[y - by][x - bx], induced[y][x]
            count -= key_v and lock_v
            if key_v ^ lock_v:
                return False
    
    return count == 0

def solution(key, lock):
    m, n = len(key), len(lock)
    induced, count = induce_lock(lock, m)
    
    comp_l = n + (m - 1)
    for _ in range(4):
        key = rotate_90_graph(key)
        for y in range(comp_l):
            for x in range(comp_l):
                if derive_key(key, induced, (x, y), count):
                    return True

    return False